omniledger/Contracts.md

```diff
@@ -0,0 +1,191 @@
+
+Given the following instruction sent by a client (some fields are omitted for
+clarity):
+
+- `InstanceID`: `[32]byte{GenesisDarc}[32]byte{0x01}`
+- `Invoke`:
+  - `Command`: `Update`
+  - `Args`: `{"Roster": NewRoster}`
+- `Signatures`: `[Sig1]`
+
+omniledger will do the following:
+
+1. find the Darc instance by looking at the first 32 bytes of the `InstanceID` given in
+the instruction, here the `GenesisDarc`, and adding 32 x 0x00 bytes
+2. create a `DarcRequest` using the `InstanceID` and the `Args`
+3. verify the request corresponds to the expression of the `Invoke_Update` rule
+in the Darc instance found in 1.
+
+## Contract Arguments
+
+A contract is immutable and has the following signature in go:
+
+```go
+type OmniLedgerContract func(coll CollectionView, tx Instruction, inCoins []Coin) (sc []StateChange, outCoins []Coin, err error)
+```
+
+Input:
+- `coll` is a read-only reference to the collection representing the global state
+of all instances
+- `tx` is the instruction sent by the client, which also holds the `InstanceID`
+pointing to the data the contract should work on
+- `inCoins` is mostly ignored for the moment, but can be used to pass around
+coins between different instructions
+
+Output:
+- `sc` is the slice of stateChanges the contract wants to apply to the global
+state. They will only be applied if all instructions in the ClientTransaction
+are valid, else they will be discarded
+- `outCoins` can be used to pass around coins
+- `err` if not nil, the contract indicates it failed, and all instructions in that
+ClientTransaction will be discarded
+
+The contract itself has access to all elements of the collection, but will mainly
+work on the data pointed to by the `tx Instruction` given as a parameter.
+
+## Instance Structure
+
+Every instance in omniledger is stored with the following information in the
+global state:
+
+- `InstanceID` is a globally unique identifier of that instance, composed of:
+  - `DarcID`, defining the access rights to that instance
+  - `Nonce`, which is randomly chosen. The special `Nonce` of `0` indicates
+the Darc responsible for all the instances starting with the same `DarcID`
+- `ContractID` points to the contract that will be called if that instance
+receives an instruction from the client
+- `Data` is interpreted by the contract and can change over time
+
+## Interaction between Instructions and Instances
+
+Every instruction sent by a client indicates the `InstanceID` it is sent to.
+Omniledger will start by verifying the authorization as described above, then
+use the `InstanceID` to look up the responsible contract for this instance and
+then send the instruction to that contract. A client can call an instance with
+one of the following three basic instructions:
+
+- `Spawn` - will ask the instance to create a new instance. The client indicates the
+requested new contract-type and the arguments. Currently only `Darc` instances can
+spawn new instances.
```

Comment:
but user defined contracts can also spawn? because that's how we expected to continue to impl eventlog.Log()